File: OneSignalSDK/onesignal/src/main/java/com/onesignal/outcomes/OSOutcomeEventsFactory.java
package com.onesignal.outcomes;

import com.onesignal.OSLogger;
import com.onesignal.OSSharedPreferences;
import com.onesignal.OneSignalAPIClient;
import com.onesignal.OneSignalDb;
import com.onesignal.outcomes.domain.OSOutcomeEventsRepository;

public class OSOutcomeEventsFactory {

    private final OSLogger logger;
    private final OSOutcomeEventsCache outcomeEventsCache;
    private final OneSignalAPIClient apiClient;
    private OSOutcomeEventsRepository repository;

    public OSOutcomeEventsFactory(OSLogger logger, OneSignalAPIClient apiClient, OneSignalDb dbHelper, OSSharedPreferences preferences) {
        this.logger = logger;
        this.apiClient = apiClient;

        outcomeEventsCache = new OSOutcomeEventsCache(logger, dbHelper, preferences);
    }

    public OSOutcomeEventsRepository getRepository() {
        if (repository == null)
            createRepository();
        else
            validateRepositoryVersion();
        return repository;
    }

    private void validateRepositoryVersion() {
        if (!outcomeEventsCache.isOutcomesV2ServiceEnabled() && repository instanceof OSOutcomeEventsV1Repository)
            return;
        if (outcomeEventsCache.isOutcomesV2ServiceEnabled() && repository instanceof OSOutcomeEventsV2Repository)
            return;

        createRepository();
    }

    private void createRepository() {
        if (outcomeEventsCache.isOutcomesV2ServiceEnabled())
            repository = new OSOutcomeEventsV2Repository(logger, outcomeEventsCache, new OSOutcomeEventsV2Service(apiClient));
        else
            repository = new OSOutcomeEventsV1Repository(logger, outcomeEventsCache, new OSOutcomeEventsV1Service(apiClient));
    }
}

File: OneSignalSDK/onesignal/src/main/java/com/onesignal/outcomes/data/OSOutcomeEventsFactory.kt
package com.onesignal.outcomes.data

import com.onesignal.OSLogger
import com.onesignal.OSSharedPreferences
import com.onesignal.OneSignalAPIClient
import com.onesignal.OneSignalDb
import com.onesignal.outcomes.domain.OSOutcomeEventsRepository

class OSOutcomeEventsFactory(private val logger: OSLogger, private val apiClient: OneSignalAPIClient, dbHelper: OneSignalDb?, preferences: OSSharedPreferences?) {
    private val outcomeEventsCache: OSOutcomeEventsCache = OSOutcomeEventsCache(logger, dbHelper!!, preferences!!)
    private var repository: OSOutcomeEventsRepository? = null

    fun getRepository(): OSOutcomeEventsRepository = if (repository != null) validateRepositoryVersion() else createRepository()

    private fun validateRepositoryVersion(): OSOutcomeEventsRepository {
        if (!outcomeEventsCache.isOutcomesV2ServiceEnabled && repository is OSOutcomeEventsV1Repository)
            return repository!!
        if (outcomeEventsCache.isOutcomesV2ServiceEnabled && repository is OSOutcomeEventsV2Repository)
            return repository!!
        return createRepository()
    }

    private fun createRepository() = if (outcomeEventsCache.isOutcomesV2ServiceEnabled)
        OSOutcomeEventsV2Repository(logger, outcomeEventsCache, OSOutcomeEventsV2Service(apiClient))
    else
        OSOutcomeEventsV1Repository(logger, outcomeEventsCache, OSOutcomeEventsV1Service(apiClient))

}